tmpl/blog/introducing-charrua-dhcp.md

```diff
@@ -0,0 +1,221 @@
+[Input](http://haesbaert.github.io/charrua-core/api/Dhcp_server.Config.html),
+which handles the input of DHCP packets.
+
+The logic is modeled in a pure functional style,
+[Dhcp_server](http://haesbaert.github.io/charrua-core/api/Dhcp_server.html) does
+not perform any IO of its own, it works by taking an input
+[packet](http://haesbaert.github.io/charrua-core/api/Dhcp_wire.html#TYPEpkt),
+ a
+ [configuration](http://haesbaert.github.io/charrua-core/api/Dhcp_server.Config.html#TYPEt)
+ and returns a possible reply to be sent by the caller, or an error/warning:
+
+#### Input
+
+```ocaml
+type result = 
+| Silence                 (* Input packet didn't belong to us, normal nop event. *)
+| Reply of Dhcp_wire.pkt  (* A reply packet to be sent on the same subnet. *)
+| Warning of string       (* An odd event, could be logged. *)
+| Error of string         (* Input packet is invalid, or some other error ocurred. *)
+
+val input_pkt : Dhcp_server.Config.t ->  Dhcp_server.Config.subnet ->
+   Dhcp_wire.pkt -> float -> result
+(** input_pkt config subnet pkt time Inputs packet pkt, the resulting action
+    should be performed by the caller, normally a Reply packet is returned and
+    must be sent on the same subnet. time is a float representing time as in
+    Unix.time or MirageOS's Clock.time. **)
+```
+
+A typical main server loop would work by:
+ 1. Reading a packet from the network.
+ 2. Unmarshaling with [Dhcp_wire.pkt_of_buf]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_wire.html#VALpkt_of_buf).
+ 3. Inputing the result with [Dhcp_server.Input.input_pkt]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt).
+ 4. Sending the reply, or logging the event from the [Dhcp_server.Input.input_pkt]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt) call.
+
+A mainloop example can be found in
+[mirage-skeleton](https://github.com/mirage/mirage-skeleton/blob/master/dhcp/unikernel.ml#L28):
+
+```ocaml
+  let input_dhcp c net config subnet buf =
+    let open Dhcp_server.Input in
+    match (Dhcp_wire.pkt_of_buf buf (Cstruct.len buf)) with
+    | `Error e -> Lwt.return (log c (red "Can't parse packet: %s" e))
+    | `Ok pkt ->
+      match (input_pkt config subnet pkt (Clock.time ())) with
+      | Silence -> Lwt.return_unit
+      | Warning w -> Lwt.return (log c (yellow "%s" w))
+      | Error e -> Lwt.return (log c (red "%s" e))
+      | Reply reply ->
+        log c (blue "Received packet %s" (Dhcp_wire.pkt_to_string pkt));
+        N.write net (Dhcp_wire.buf_of_pkt reply)
+        >>= fun () ->
+        log c (blue "Sent reply packet %s" (Dhcp_wire.pkt_to_string reply));
+        Lwt.return_unit
+```
+
+As stated,
+[Dhcp_server.Input.input_pkt](http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt)
+does not perform any IO of its own, it deals only with the logic of inputing a
+DHCP packet and returning a possible answer, which should then be sent by the
```

Comment:
I find this wording a bit confusing.  Wouldn't `Dhcp_server.Input.response` be a better name with the description being “[it] does not perform any IO of its own, it only deals with the logic of *analyzing* a DHCP packet and *building* a possible answer,...**”?